core/model.py

```diff
@@ -0,0 +1,220 @@
+            self._database.update(self._get_collection_name(), self._id, self._get_data_store_object())
+        else:
+            self.created_at = current_utc_time
+            saved_data_id = self._database.create(self._get_collection_name(), self._get_data_store_object())
+            self._id = saved_data_id
+        return self
+
+    def update(self: T, data: dict) -> T:
+        for key in data:
+            if key in self.__annotations__:
+                setattr(self, key, data[key])
+        self.save()
+        return self
+
+    @classmethod
+    def _get_collection_name(cls: T):
+        object_name = cls.__name__
+        snake_case_name = re.sub(r'(?<!^)(?=[A-Z])', '_', object_name).lower()
+        return snake_case_name
+    
+    @classmethod
+    def get(cls: T, id: Union[str, ObjectId]) -> T:
+        class_instance = cls()
+        return cls(class_instance._get_data_load_object(class_instance._database.get(cls._get_collection_name(), id)))
+
+    @classmethod
+    def query(cls: T, query: dict, sort=None) -> list[T]:
+        class_instance = cls()
+        return [cls(class_instance._get_data_load_object(record)) for record in class_instance._database.query(cls._get_collection_name(), query, sort)]
+
+    def unlink(self):
+        '''
+        Remove record from the database
+        '''
+        self._database.delete(self._get_collection_name(), self._id)
+        self._id = None
+
+    def dict(self) -> dict:
+        '''
+        Return data object as dictionary. Can be used for serialization. 
+        '''
+        data_as_dict = {}
+        
+        for key in self._fields:
+            value = getattr(self, key)
+            data_as_dict[key] = self._serialize(value)
+        return data_as_dict
+
+    def parse_datetime(self, format="%Y-%m-%d"):
+        '''
+        Update datetime from request in the model
+        '''
+        for field in self._fields:
+            if hasattr(self, field) and self.__annotations__.get(field) is datetime and isinstance(getattr(self, field), str):
+                datetime_object = datetime.strptime(getattr(self, field), format)
+                setattr(self, field, datetime_object)
+
+    def _serialize(self, value):
+        serialized_data = None
+        if isinstance(value, ObjectId):
+            serialized_data = str(value)
+        elif isinstance(value, datetime):
+            serialized_data = str(value)
+        elif isinstance(value, Model):
+            serialized_data = value.dict()
+        elif isinstance(value, Enum):
+            serialized_data = value.value
+        elif isinstance(value, list):
+            serialized_data = []
+            for each_value in value:
+                serialized_data.append(self._serialize(each_value))
+        else:
+            serialized_data = value
+        return serialized_data
+
+    @classmethod
+    def _get_flask_restx_field(cls, type_of_field):
+        is_list = len(get_args(type_of_field)) == 1
```

Comment:
Why this can't be `type_of_field is list` like every other type?

Reply:
`type_of_field` has a type of `type` not `list`. So we cannot check `type_of_field is list`. This will return false.

Reply:
I see. Amazingly horrible ass-backwards solution from the python team...